### PR TITLE
SDG_2: No need for thread local storage as readonly

### PR DESCRIPTION
**Pull request by afabri**
Solves #1403

### CODE REVIEW

File: Segment_Delaunay_graph_2/include/CGAL/Segment_Delaunay_graph_2.h [hunk 1]
```diff
@@ -44,13 +44,15 @@
...
 #include <CGAL/Segment_Delaunay_graph_2/Constructions_C2.h>
 
-#include <CGAL/Iterator_project.h>
 #include <CGAL/utility.h>
 
 #include <CGAL/spatial_sort.h>
 #include <CGAL/Spatial_sort_traits_adapter_2.h>
 
+#include <CGAL/tss.h>
+
```
Comment (sloriot) [inline Segment_Delaunay_graph_2/include/CGAL/Segment_Delaunay_graph_2.h]: The include directive added is not needed I think

**Comment by sloriot** [comment]
@afabri I've marked the point solved by this PR in the issue. Unless I'm wrong you haven't fixed them all.

File: Segment_Delaunay_graph_2/include/CGAL/Segment_Delaunay_graph_2.h (hunk 1 above)
Comment (afabri) [inline Segment_Delaunay_graph_2/include/CGAL/Segment_Delaunay_graph_2.h]: For void `print_error_message(const Tag_false&) const`

**Comment by afabri** [comment]
@sloriot  this seems ready to me; please check

File: Segment_Delaunay_graph_2/include/CGAL/Segment_Delaunay_graph_2/Segment_Delaunay_graph_2_impl.h [hunk 2]
```diff
@@ -901,17 +901,7 @@ insert_intersecting_segment_with_tag(const Storage_site_2& /* ss */,
...
-  // hack to avoid nasty warning for G++ 3.4 on Darwin
-  static int i;
-#else
-  static int i = 0;
-#endif
-  if ( i == 0 ) {
-    i = 1;
-    print_error_message();
-  }
+  CGAL_assertion_msg(false, "segment segment intersection in the input");
```
Comment (lrineau) [inline Segment_Delaunay_graph_2/include/CGAL/Segment_Delaunay_graph_2/Segment_Delaunay_graph_2_impl.h]: Should be:
```C++
CGAL_error_msg("segment segment intersection in the input");
```

File: Segment_Delaunay_graph_2/include/CGAL/Segment_Delaunay_graph_2/Triangulation_face_base_with_edges_2.h [hunk 3]
```diff
@@ -125,18 +125,18 @@ class Triangulation_face_base_with_edges_2
...
+    return Edge(Face_handle(), sentinel_index());
   }
 
   // initialization of in-place list pointers
   void init() {
     for (int i = 0; i < 3; i++) {
-      next_edge_in_list[i] = sentinel_edge();
-      prev_edge_in_list[i] = sentinel_edge();
+      next_edge_in_list[i] = Edge(Face_handle(),-1);
+      prev_edge_in_list[i] = Edge(Face_handle(),-1);
```
Comment (lrineau) [inline Segment_Delaunay_graph_2/include/CGAL/Segment_Delaunay_graph_2/Triangulation_face_base_with_edges_2.h]: Why this change? You have already changed `sentinel_edge()`. Let's not hardcode the value `-1` at multiple occurrences in the code.

File: Segment_Delaunay_graph_2/include/CGAL/Segment_Delaunay_graph_face_base_2.h [hunk 4]
```diff
@@ -64,7 +64,7 @@ class Segment_Delaunay_graph_face_base_2
     // method to initialize the in-place edge list
     void initialize_in_place_edge_list()
     {
-      static Edge SENTINEL_QUEUE_EDGE = Edge(Face_handle(), -1);
+      static const Edge SENTINEL_QUEUE_EDGE = Edge(Face_handle(), -1);
```
Comment (lrineau) [inline Segment_Delaunay_graph_2/include/CGAL/Segment_Delaunay_graph_face_base_2.h]: Not consistent with the changes made to `sentinel_edge()`, above.

File: Segment_Delaunay_graph_2/include/CGAL/Segment_Delaunay_graph_hierarchy_2.h [hunk 5]
```diff
@@ -367,17 +364,7 @@ class Segment_Delaunay_graph_hierarchy_2
...
-    // hack to avoid nasty warning for G++ 3.4 on Darwin
-    static int i;
-#else
-    static int i = 0;
-#endif
-    if ( i == 0 ) {
-      i = 1;
-      print_error_message();
-    }
+    CGAL_assertion_msg(false, "SDG::Insert aborted: intersecting segments found");
```
Comment (lrineau) [inline Segment_Delaunay_graph_2/include/CGAL/Segment_Delaunay_graph_hierarchy_2.h]: `CGAL_error_msg`

File: Segment_Delaunay_graph_2/include/CGAL/Segment_Delaunay_graph_2/in_place_edge_list.h [hunk 6]
```diff
@@ -243,7 +243,7 @@ class In_place_edge_list_for_sdg {
 
   void remove(const Edge& e) {
     CGAL_precondition( is_in_list(e) );
-    static Edge SENTINEL_QUEUE_EDGE = Edge(Face_handle(), -1);
+    static const Edge SENTINEL_QUEUE_EDGE = Edge(Face_handle(), -1);
```
Comment (lrineau) [inline Segment_Delaunay_graph_2/include/CGAL/Segment_Delaunay_graph_2/in_place_edge_list.h]: I do not think that this initialization of a `static const` variable in a scope is guaranteed thread-safe, before C++11.

Comment (afabri) [inline Segment_Delaunay_graph_2/include/CGAL/Segment_Delaunay_graph_2/in_place_edge_list.h]: I removed the word static.

File: Segment_Delaunay_graph_2/include/CGAL/Segment_Delaunay_graph_2.h [hunk 7]
```diff
@@ -1427,13 +1425,7 @@ class Segment_Delaunay_graph_2
...
   void print_error_message(const Tag_false&) const
   {
-    static int i = 0;
-
-    if ( i == 0 ) {
-      i++;
-      std::cerr << "SDG::Insert aborted: intersecting segments found"
-		<< std::endl;
-    }
+    CGAL_assertion_msg(false, "SDG::Insert aborted: intersecting segments found");
```
Comment (sloriot) [inline Segment_Delaunay_graph_2/include/CGAL/Segment_Delaunay_graph_2.h]: as Laurent already pointed out, you can use `CGAL_error_msg`

**Comment by lrineau** [comment]
SDG_2 and SDG_Linf_2 are fully red. Probably due to this PR:
https://cgal.geometryfactory.com/CGAL/testsuite/results-4.10-Ic-15.shtml#Segment_Delaunay_graph_2

Could you try locally?

**Comment by afabri** [comment]
The tests had failed as I turned a message into an error. It is a message again, so it could be tested again.
